datafusion/expr/src/type_coercion/binary.rs

```diff
@@ -238,13 +238,32 @@ fn comparison_binary_numeric_coercion(
         (_, Decimal128(_, _)) => get_comparison_common_decimal_type(rhs_type, lhs_type),
         (Float64, _) | (_, Float64) => Some(Float64),
         (_, Float32) | (Float32, _) => Some(Float32),
-        (Int64, _) | (_, Int64) => Some(Int64),
-        (Int32, _) | (_, Int32) => Some(Int32),
-        (Int16, _) | (_, Int16) => Some(Int16),
-        (Int8, _) | (_, Int8) => Some(Int8),
+        // start checking from Int64, that is the most inclusive integer type
```

Comment:
Could we express the same logic more concisely with something like:

```
      (Int64, _) | (_, Int64) => Some(Int64),
        (Int32, _) | (_, Int32) => Some(Int64),
        (Int16, _) | (_, Int16) => Some(Int32),
        (Int8, _) | (_, Int8) => Some(Int16),
```

Reply:
Likewise I think the same pattern could be used for `UInt*` variants

Reply:
Wouldn't that map (`Int32`, `Int16`) to `Int64`? I think @berkaysynnada's intent is to map to the narrowest correct type, which in this case would be `Int32`.

I am not sure if the `match` pattern is the most succinct representation of the table in the PR text though.

Reply:
> Wouldn't that map (Int32, Int16) to Int64? 

Yes.

I guess I was thinking of the more general case for arithmetic where  `i32::MAX + i16::MAX` requires a `i64` to store

Reply:
I am confused by the sum. Aren't we just "merging" the schemas? In that case, the narrowest type that can represent both sides should be chosen, no? Am I missing something here?

Reply:
I think I am just confused 

For example this PR adds this rule:

```
(Int8, UInt16) => Some(Int32),
```

Which isn't just merging the schema (e.g. UInt16) in my mind, but is extending it in some way. In this case, so that the entire range of `Int8` and `UInt16` can be represented -- which now makes sense

I think what would help me would be:
1. Add some comments explaining the rationale for these rules
2. (maybe) add test coverage (ideally looking like your beautiful table from the PR description) that helps to illustrate the point

However, I don't think this is needed prior to merging this PR

Reply:
Just added comments to explain the reasoning of the `match` patterns. Also took note to add some tests in a follow-on PR. Thanks for the review!